sams-scala-tutorial/serialization-exceptions-and-memory-leaks-no-ws.md

```diff
@@ -13,14 +13,147 @@ Talk:
+     * In both these 2 examples, closures have a reference outside respectively(unSerializableClassHandler and unSerializableObjectHandler).
+     * Obviously, because they do not extend Serializable, neither of them can be serialized. 
+     * */
+
+    /*How to solve this kind of exceptions? */
+
+    /*
+     * solution 1: Use reference outside and make object or class extend interface Serializable
+     * In this way, outer reference can be serialized by colusure cleaner and the task can be shipped from driver to data nodes.
+     * */
+    val SerializableClassHandler = new SerializableClass(1)
+    rdd.map { x => SerializableClassHandler.func_1 }.count()
+
+    val SerializableObjectHandler = SerilizableObject
+    rdd.map { x => SerializableObjectHandler.func_1(x) }.count()
+    /*
+     * Above 2 examples can be executed successfully because SerializableClassHandler and SerializableObjectHandler can be serialized by closure cleaner.
+     * */
+
+    /*
+     * Solution 2: Do not use outside reference, and instantiate a object for every element in RDD.
+     * In this way, for each element in RDD, a class will be instantiated or a local object will be referenced in its own node, so there is no worry about serialization.
+     * */
+    rdd.map { x =>
+      {
+        val handler = new NonSerializableClass(1)
+        handler.func_1
+      }
+    }.count()
+
+    rdd.map { x =>
+      {
+        val handler = NonSerilizableObject
+        handler.func_1(x)
+      }
+    }.count()
+    /*
+     * Above 2 examples works because no objects need to be serialized. 
+     * As I use nonSerializable object and class in RDD closure, whether serializable or not does not matter. 
+     * */
+
+    /*
+     * Solution 3: Do not use outside reference, and only use static functions.
+     * In scala, static functions only exist in Object so this method only work for Object
+     * */
+    rdd.map { x =>
+      {
+        NonSerilizableObject.func_1(x)
+      }
+    }.count()
+    /*
+     * The example works. Actually this method is the same as second example in solution 2.
+     * Method func_1 is directly from object without any outside reference.
+     * */
+
+    /*
+     * Tip for object in spark
+     * A static object is bound to one JVM, so if a static variable is used in RDD closure, it will only change within its own node
+     * */
+    rdd.map { x =>
+      {
+        val handler = staticObj
+        handler.counter += 1
+        (x, handler.counter)
+      }
+    }.collect().foreach { println }
+    println(staticObj.counter)
+    /* results:
+     * (1,3)
+     * (2,2)
+     * (3,2)
+     * (4,3)
+     * (5,2)
+     * 1
+     * After counter increment in worker node, the counter in driver node's object is still 1.
+     * That means worker node have objects attached itself. If you want to change values in driver node, the variable needs to be broadcasted.
+     * */
+  }
+}
+
```

Comment:
@QingyangKong Would be nicer to put these at the top of the example so audience has shorter distance to scroll.

Can make the definitions more terse by getting rid of `{}` for single line expressions (which is a good practice as it improves compiler checking) and putting on one line.